### PR TITLE
Set system to sleep on basis of PR#551

### PR DESCRIPTION
**Pull request by Gemba**
Evolution of initial #550 and @jrassa scripting support #551.

-  Configurable system sleep time (0 for off), will call out to `<scriptdir>/sleep/<scripts>` when timeout reached. Will call `<scriptdir>/wake/<scripts>` on user interaction.
-  Additional scripting events `screensaver-start` and `screensaver-stop`. See also: https://github.com/jrassa/EmulationStation/wiki/Scripting
-  Linux specific: Runs executable (`u+x`) scripts only
-  Fixed indention (character) according to majority of indention characters used in source file.

### CODE REVIEW

**Comment by joolswills** [comment]
Thanks for this. My only comment from an initial look would be that it would be better to have the majority of cosmetic fixes (indent) done in a separate commit. Makes for easier reading / management of the change.

**Comment by Gemba** [comment]
Hi @joolswills and maintainers, have been away a while. Now I updated the PR and kept the pristine whitespaces, bear with me for the very few whitespace changes still in there.

**Comment by Gemba** [comment]
Bummer! Hit "Comment" instead of "Reopen and comment".

**Comment by theofficialgman** [comment]
@joolswills @Gemba Is there any reason that this has not been merged? I was looking into this today and based on the wiki I thought that it had only to figure out that the changes have not been merged. Trying to use the screensaver-start and screensaver-stop scripting features which will be awesome.

**Comment by Gemba** [comment]
I have no information on the merge plan, @theofficialgman. I can imagine other topics are more important at the moment (e.g. Retropie at RPi4).

Thanks for the feedback that this PR works as expected. :-)

File: es-core/src/Window.cpp [hunk 1]
```diff
@@ -269,16 +269,13 @@ void Window::render()
...
 	{
-		if (!isProcessing() && mAllowSleep && (!mScreenSaver || mScreenSaver->allowSleep()))
-		{
-			// go to sleep
-			if (mSleeping == false) {
-				mSleeping = true;
-				onSleep();
-			}
+		unsigned int systemSleepTime = (unsigned int)Settings::getInstance()->getInt("SystemSleepTime");
+		if(!isProcessing() && mAllowSleep && systemSleepTime != 0 && mTimeSinceLastInput >= systemSleepTime) {
```
Comment (pjft) [inline es-core/src/Window.cpp:276]: Apologies if it's somewhere else, but it's not clear to me why are removing "(!mScreenSaver || mScreenSaver->allowSleep())" ?
If I read it correctly, those are there for ES to only go to sleep if either the screensaver isn't active, or if the screensaver allows sleeping. Won't this change the behavior and always force sleep over the screensaver?
Maybe that is intended, but wanted to probe into this.

**Comment by pjft** [comment]
From reading the code, I have few questions about it. Since the entire events logic was implemented by @jrassa , I'd definitely appreciate his review of the changes here as it's not an area of the code I'm familiar with.

**Comment by hex007** [comment]
You might also want to test out the compatibility with various PowerSaver modes. I can review that part of the logic of needed.

File: es-core/src/Window.cpp (hunk 1 above)
Comment (Gemba) [reply es-core/src/Window.cpp:276]: D'oh. Somehow my comment for resolving did not make it through. Retyping it again.

The initial confusion may be caused by the fact that I pm'ed jdrassa about the sleep/wakeup events for the wiki page (Aug. 2019). Thus it looked like it was completly implemented, but it wasn't.

For your challenge in this code segment: Good to have strict reviews. It was intented to enter sleep without having an screensaver set.
If you have a screensaver enabled it is not possible via UI to set sleep timeout < screensaver timeout. The user will get a note and in turn the timeout is set that sleep and screensaver have the same value. However, it is possible to force sleep timeout < screensaver timeout via the command line, but this is a different story: I assume someone knows what he/she is doing when editing files manually.

You can set sleep timeout to 0 which has the same semantic as in the screensaver (=off).

Comment (pjft) [reply es-core/src/Window.cpp:276]: Thanks!

**Comment by jrassa** [comment]
Thanks and sorry for the delay.

**Comment by Gemba** [comment]
No worries. Thanks, @jrassa 

I assume issue #511 can be closed then.

@hex007 can we elaborate this in the "Ideas and Development" retropie forum first pls?

**Comment by hex007** [comment]
@Gemba sure. Please tag me in your post. Thanks for the heads up.

**Comment by theofficialgman** [comment]
its been a while since I updated emulationstation and I was still using my fork with this PR from Apr 26, 2020

I updated to emulationstation dev version today and it seems that the scripts don't ever launch.
Has the functionality changed since then? I am creating my scripts in ~/.emulationstation/scripts/screensaver-start as told by this wiki page https://github.com/jrassa/EmulationStation/wiki/Scripting

**Comment by theofficialgman** [comment]
I checked the es_log.txt and I get this error even though my script is indeed exectuable and I have done chmod u+x to verify as well
```
Apr 18 21:37:46 lvl1: 	/home/garrett/.emulationstation/scripts/screensaver-start/create_file is not executable. Did you 'chmod u+x'?. Skipping this script.```

**Comment by theofficialgman** [comment]
alright so this does not appear to be a problem with this PR but actually something introduced after
originally the isExecutable check was written as 
```
#ifndef WIN32 // osx / linux
		bool isExecutable(const std::string& _path) {
			struct stat64 st;
			if(stat64(_path.c_str(), &st) == 0){
				mode_t perm = st.st_mode;
				// regular files and executables but not setuid, setgid, shared text (mode 0755)
				mode_t mask = S_IFREG | S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH;
				if(perm & mask == perm){
					return true;
				}
			}
			return false;
		} // isExecutable
#endif
```
but now it is written as 
```
#if !defined(_WIN32)
		bool isExecutable(const std::string& _path)
		{
			const std::string path = getGenericPath(_path);

			// regular files and executables but not setuid, setgid, shared text (mode 0755)
			const mode_t  mask = S_IFREG | S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH;
			struct stat64 info;

			// check if stat64 succeeded
			if(stat64(path.c_str(), &info) != 0)
				return false;

			// check for mask attributes only
			return ((info.st_mode & mask) == info.st_mode);

		} // isExecutable
#endif // !_WIN32
```

**Comment by cmitu** [comment]
@theofficialgman should be handled by #732.

**Comment by Gemba** [comment]
@theofficialgman the former and the latter implementation do only return true for 755 file mode.
The latter sets the evaluation precedence right by using brackets in `(info.st_mode & mask) == info.st_mode`. Other than that the implementations are semantically the same.
However, if it would have been written as `(info.st_mode & mask) == mask` it would have allowed files with 755 permissions or more permission bits (e.g. 775) to pass.
As @cmitu  mentioned, it is fixed in the PR #732, after we concluded what would be a more relaxed check on the permissions.

**Comment by tomaz82** [comment]
Thanks @Gemba for clearing that up, I have been looking at the old code, and my new code, and could not find anything different with it. Turns out there wasn't anything :P

**Pull request merged**